### PR TITLE
Avoid panics on error while encoding/decoding ListValue::Array as protobuf

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Follow on to https://github.com/apache/arrow-datafusion/pull/7629

## Rationale for this change

I noticed some panics while reviewing https://github.com/apache/arrow-datafusion/pull/7629 from @jayzhan211  but I didn't want to delay merging that PR any longer.

## What changes are included in this PR?

This commit avoids panics in the error handling in the ScalarValue::List protobuf encoding by using the `?` operator
instead unwrap. It also removes some unused variants that are no longer used after
https://github.com/apache/arrow-datafusion/pull/7629

## Are these changes tested?

By existing tests and the rust compiler

## Are there any user-facing changes?
Less panics on bad data

### CODE REVIEW

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 1]
```diff
@@ -655,29 +656,36 @@ impl TryFrom<&protobuf::ScalarValue> for ScalarValue {
...
 
-                let message = root_as_message(ipc_message.as_slice()).unwrap();
+                let message = root_as_message(ipc_message.as_slice()).map_err(|e| {
+                    Error::General(format!(
+                        "Error IPC message while deserializing ScalarValue::List: {e}"
+                    ))
+                })?;
+                let buffer = Buffer::from(arrow_data);
 
-                // TODO: Add comment to why adding 0 before arrow_data.
```
Comment (alamb) [inline datafusion/proto/src/logical_plan/from_proto.rs:663]: The 0 was added in the arrow.rs code as it was a test for unaligned data. The zero is not needed for this cod

cc @jayzhan211

File: datafusion/proto/src/logical_plan/to_proto.rs [hunk 2]
```diff
@@ -56,13 +56,6 @@ use datafusion_expr::{
 pub enum Error {
     General(String),
 
-    InconsistentListTyping(DataType, DataType),
```
Comment (alamb) [inline datafusion/proto/src/logical_plan/to_proto.rs:59]: These variants are no longer used

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 3]
```diff
@@ -655,29 +656,36 @@ impl TryFrom<&protobuf::ScalarValue> for ScalarValue {
                 let schema: Schema = if let Some(schema_ref) = schema {
                     schema_ref.try_into()?
                 } else {
-                    return Err(Error::General("Unexpected schema".to_string()));
+                    return Err(Error::General(
```
Comment (comphead) [inline datafusion/proto/src/logical_plan/from_proto.rs:659]: Should we return some of Datafusion Errors instead? Maybe IoError or InternalError?

**Review by jayzhan211 — APPROVED** [review]
LGTM. Thanks, @alamb !

File: datafusion/proto/src/logical_plan/from_proto.rs (hunk 3 above)
Comment (alamb) [reply datafusion/proto/src/logical_plan/from_proto.rs:659]: I am not sure why the from_proto and to_proto modules have their own error types and don't use `DataFusionError` (or even wrap `DataFusionError`)

It does seem like that would be an improvement. Maybe I can file a follow on ticket to do so.

**Review by Dandandan — APPROVED** [review]

**Pull request merged**